### PR TITLE
Lava quad

### PR DESCRIPTION
**Pull request by St4rman**
Finally creating a pull request for this. 
- Lava Quad in , a little off when you look at it from far away, the texture is a bit blurry 
- Added a end Portal functionality but that is commented out in this branch, I will do that in a seperate branch to keep this clean

### CODE REVIEW

**Review by Odycious — APPROVED** [review]
Changes look good to me!

File: Client/Renderer.cpp [hunk 1]
```diff
@@ -21,7 +21,7 @@ GameTechRenderer::GameTechRenderer(GameWorld& world, Canvas& canvas) : OGLRender
     debugShader  = new OGLShader("debug.vert", "debug.frag");
     shadowShader = new OGLShader("shadow.vert", "shadow.frag");
     textShader = std::make_shared<OGLShader>("text.vert", "text.frag");
-    defaultShader = new OGLShader("scene.vert", "Buffer.frag");
```
Comment (Joerseven) [inline Client/Renderer.cpp]: Why has their been a whole shader change for deferred rendering?

File: Client/Renderer.cpp [hunk 2]
```diff
@@ -602,6 +601,8 @@ void GameTechRenderer::RenderSkybox() {
 }
 
 void GameTechRenderer::RenderCamera() {
+    glEnable(GL_BLEND);
+    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
```
Comment (Joerseven) [inline Client/Renderer.cpp]: Deferred rendering doesn't support alpha blending not sure why this has been added.

File: Assets/Shaders/scene.frag [hunk 3]
```diff
@@ -12,27 +13,37 @@ uniform vec3	cameraPos;
...
-{
+void main(void)	{
 	float shadow = 1.0; // New !
+
+	mat3 TBN = mat3(normalize(IN.tangent), normalize(IN.binormal), normalize(IN.normal));
 	
+	vec3 bumpNormal = texture2D(bumpTex, IN.texCoord).rgb * 2.0 - 1.0;
+	bumpNormal = normalize(TBN * normalize(bumpNormal));
+
+	fragColour[1] = vec4(bumpNormal.xyz * 0.5 + 0.5, 1.0);
```
Comment (Joerseven) [inline Assets/Shaders/scene.frag:37]: This is overwritten a few lines down.

**Review by Joerseven — DISMISSED** [review]
This PR is like a hot dude but he still gives you a  bit of the ick, know what I mean? Good looking, you just know that there's a red flag hidden away somewhere.

File: Client/Renderer.cpp (hunk 1 above)
Comment (St4rman) [reply Client/Renderer.cpp]: Acc to matthew:  in his graphics coursework he had his def rendering this fragment shader so he had that in the project. now that hes put it in scene.frag this shouldbe ok to remove

File: Client/Renderer.cpp (hunk 2 above)
Comment (St4rman) [reply Client/Renderer.cpp]: Removed. Added them before deferred to check some stuff if i had to guess. Not entirely sure however.

File: Assets/Shaders/scene.frag (hunk 3 above)
Comment (St4rman) [reply Assets/Shaders/scene.frag:37]: this i dunno

**Comment by Joerseven** [comment]
WRONG

File: Client/States/GameplayState.cpp [hunk 4]
```diff
@@ -725,3 +730,22 @@ float GameplayState::CalculateCompletion(Vector3 playerCurPos){
...
+    AddLava(lavaPos + Vector3(0, 5, 0));
+    AddEndPortal(endPos);
+
+}
+
+void GameplayState::AddLava(Vector3 position){
+    auto LavaQuad = new GameObject();
+    LavaQuad->GetTransform().SetOrientation(Quaternion(Matrix4::Rotation(90, {1,0,0})));
+    replicated->AddTestObjectToLevel(LavaQuad, *world, {1000,1000,1000}, position, false);
+    LavaQuad->SetRenderObject(new RenderObject(&LavaQuad->GetTransform(), resources->GetMesh("QuadSequel.msh"), resources->GetTexture("VorDef.png"), resources->GetShader("lava")));
```
Comment (DelCherub) [inline Client/States/GameplayState.cpp]: We should change this back to normal quad lmao

File: Assets/Meshes/QuadSequel.msh [hunk 5]
```diff
@@ -0,0 +1,33 @@
+MeshGeometry
```
Comment (DelCherub) [inline Assets/Meshes/QuadSequel.msh]: delet this

File: Client/States/GameplayState.cpp [hunk 6]
```diff
@@ -725,3 +730,22 @@ float GameplayState::CalculateCompletion(Vector3 playerCurPos){
     auto progress = playerCurPos - startPos;
     return progress.Length()/levelLen;
 }
+
+void GameplayState::RenderFlairObjects(){
```
Comment (DelCherub) [inline Client/States/GameplayState.cpp:993]: What is a flair object? Like just a decorative one?

**Review by DelCherub — CHANGES_REQUESTED** [review]
Looks good, just change the quad mesh please
doge

File: Client/States/GameplayState.cpp (hunk 4 above)
Comment (St4rman) [reply Client/States/GameplayState.cpp]: deleted yes yes killed deleted

File: Assets/Meshes/QuadSequel.msh (hunk 5 above)
Comment (St4rman) [reply Assets/Meshes/QuadSequel.msh]: killed

File: Client/States/GameplayState.cpp (hunk 6 above)
Comment (St4rman) [reply Client/States/GameplayState.cpp:993]: Yeah, just objects that are decor. This is what would be used if people needed to add like more just flair objects like the portal and all. Just so initlevel isnt like a massive method

**Review by DelCherub — APPROVED** [review]
Lavaquad is sooooooo awesome
image

**Comment by AdSand** [comment]
Lava quad could be it's own game at this point

**Review by Odycious — APPROVED** [review]

**Pull request merged**